gitzones/auth

```diff
@@ -0,0 +1,167 @@
+        except FileNotFoundError:
+            pass
+
+    def add(self, proto: str, key: str, label: str) -> None:
+        """ Adds an authorized key. """
+        if label in self.db:
+            raise AuthorizedKeysException(f'Key with label "{label}" already exists.')
+        self.db[label] = proto + " " + key
+        self._save()
+
+    def rm(self, label: str) -> None:
+        """ Removes an authorized key, identified by its label. Raises if key with given label cannot be found. """
+        try:
+            del self.db[label]
+        except KeyError:
+            raise AuthorizedKeysException(
+                f'Could not find authorized key with label "{label}".'
+            )
+        self._save()
+
+    def ls(self) -> Dict[str, str]:
+        """ Returns dictionary of authorized keys, identified by their labels. """
+        return self.db.copy()
+
+    def _save(self) -> None:
+        with open(self.location, "w") as f:
+            for label, proto_key in self.db.items():
+                f.write(proto_key + " " + label + "\n")
+
+
+def main(args: List[str]) -> None:
+    """ Command line application main entry point. """
+    parser = argparse.ArgumentParser(
+        usage=usage,
+    )
+    parser.add_argument("command", help="Subcommand to run")
+    parsed = parser.parse_args(args[0:1])
+
+    cmd = {
+        "id": cmd_id,
+        "add": cmd_add,
+        "rm": cmd_rm,
+        "ls": cmd_ls,
+        "cp": cmd_cp,
+    }.get(parsed.command, None)
+    if not callable(cmd):
+        print(f'Unrecognized command "{parsed.command}"')
+        parser.print_usage()
+        exit(1)
+
+    try:
+        cmd(args[1:])
+    except AuthorizedKeysException as e:
+        sys.stderr.write(str(e) + "\n")
+        exit(1)
+
+
+def cmd_id(args: List[str]) -> None:
+    """ Entrypoint for CLI command that shows this hosts SSH public key. """
+    parser = argparse.ArgumentParser()
+    parser.parse_args(args)
+    with open(host_public_key_file, "r") as f:
+        print(f'desec.{os.environ["DESECSTACK_DOMAIN"]} {f.readline().strip()}')
+
+
+def cmd_add(args: List[str]) -> None:
+    """ Entrypoint for CLI command that adds an authorized SSH key. """
+    parser = argparse.ArgumentParser()
+    parser.add_argument("proto", help="Protocol used with given key.")
+    parser.add_argument("key", help="Public key of authorized key pair.")
+    parser.add_argument("label", help="Label under which the key is stored.", nargs="+")
+    parsed = parser.parse_args(args)
+    AuthorizedKeys().add(parsed.proto, parsed.key, " ".join(parsed.label))
+
+
+def cmd_rm(args: List[str]) -> None:
+    """ Entrypoint for CLI command that removes an authorized SSH key. """
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "label", help="All keys with this label will be removed.", nargs="+"
+    )
+    parsed = parser.parse_args(args)
+    AuthorizedKeys().rm(" ".join(parsed.label))
```

Comment:
Why the `join`?

Reply:
so that labels can contain spaces

Reply:
Ah, I had missed the `nargs="+"`. I guess that's one way. On the other hand, wouldn't we expect the user to enclose labels in quotes `"` if they contain whitespace (just like for any other shell argument)?

I'm pointing this out because it's a deviation from how shell arguments with whitespace are usually handled. Maybe I'm getting something wrong here ...

Reply:
nah you're right, I just thought it's a cheap way to do it. I can remove that?

Reply:
I think you can -- what do you mean?

Reply:
I removed it, now it behaves like a shell application should